demos/bubble_shear.py

```diff
@@ -0,0 +1,483 @@
+# algorithms, where the mesh is iteratively adapted at the end of the simulation.
+#
+# We begin by defining the advection problem. We consider the advection equation
+#
+# .. math::
+#   \begin{array}{rl}
+#       \frac{\partial c}{\partial t} + \nabla\cdot(\mathbf{u}c)=0& \text{in}\:\Omega,\\
+#       c=0 & \text{on}\:\partial\Omega,\\
+#       c=c_0(x,y) & \text{at}\:t=0,
+#   \end{array},
+#
+# where :math:`c=c(x,y,t)` is the sought tracer concentration,
+# :math:`\mathbf{u}=\mathbf{u}(x,y,t)` is the background velocity field, and
+# :math:`\Omega=[0, 1]^2` is the spatial domain of interest with boundary
+# :math:`\partial\Omega`.
+#
+# The background velocity field :math:`\mathbf{u}(x, y, t)` is chosen to be periodic in
+# time, and is given by
+#
+# .. math::
+#   \mathbf{u}(x, y, t) := \left(2\sin^2(\pi x)\sin(2\pi y), -\sin(2\pi x)\sin^2(\pi y) \right) \cos(2\pi t/T),
+#
+# where :math:`T` is the period. Note that the velocity field is solenoidal. At each
+# timestep of the simulation we will update this field so we define a function that will
+# return its vector expression. ::
+
+from firedrake import *
+
+T = 6.0
+
+
+def velocity_expression(mesh, t):
+    x, y = SpatialCoordinate(mesh)
+    u_expr = as_vector(
+        [
+            2 * sin(pi * x) ** 2 * sin(2 * pi * y) * cos(2 * pi * t / T),
+            -sin(2 * pi * x) * sin(pi * y) ** 2 * cos(2 * pi * t / T),
+        ]
+    )
+    return u_expr
+
+
+# We proceed similarly with prescribing initial conditions. At :math:`t=0`, we
+# initialise the tracer concentration :math:`c_0 = c(x, y, 0)` to be :math:`1` inside
+# a circular region of radius :math:`r_0=0.15` centred at :math:`(x_0, y_0)=(0.5, 0.65)`
+# and :math:`0` elsewhere in the domain. Note that such a discontinuous function will
+# not be represented well on a coarse uniform mesh. ::
+
+
+def get_initial_condition(mesh):
+    x, y = SpatialCoordinate(mesh)
+    ball_r, ball_x0, ball_y0 = 0.15, 0.5, 0.65
+    r = sqrt(pow(x - ball_x0, 2) + pow(y - ball_y0, 2))
+    c0 = Function(FunctionSpace(mesh, "CG", 1))
+    c0.interpolate(conditional(r < ball_r, 1.0, 0.0))
+    # return conditional(r < ball_r0, 1.0, 0.0)
+    return c0
+
+
+# Now we are ready to solve the advection problem. Since we will solve the problem
+# several times, we will wrap up the code in a function that we can easily call
+# for each subinterval. The function takes the mesh over which to solve the problem,
+# time interval :math:`[t_{\text{start}}, t_{\text{end}}]` over which to solve it, and
+# the initial condition :math:`c_0 = c(x, y, t_{\text{start}})`. The function returns
+# the solution :math:`c(x, y, t_{\text{end}})`.
+#
+# Note that we must include streamline upwind Petrov Galerkin (SUPG) stabilisation in
+# the test function in order to ensure numerical stability. ::
+
+# Time-stepping parameters
+# dt = 0.01  # timestep size
+# theta = 0.5  # Crank-Nicolson implicitness
```

Comment:
Sorry, forgot to delete this! Done in 118de05d0291868951e7469d53a715764e30dc5f

I was testing having `dt` and `theta` as floats like this, rather than functions of R, as in the code. With this change, the uniform results stay the same, while the classical and metric advection results change. I'll ask you or Stephan at some point why this is the case :)

Reply:
Perhaps open an issue labelled with `question`?

Reply:
Opened #150